modules/xmpp/RtxModifier.js

```diff
@@ -0,0 +1,276 @@
+    let primarySsrcCname = primarySsrcInfo.cname;
+
+    let previousAssociatedRtxStream = 
+        getRtxSsrc (videoMLine, primarySsrc);
+    if (previousAssociatedRtxStream === rtxSsrc) {
+        logger.info(rtxSsrc + " was already associated with " +
+            primarySsrc);
+        return;
+    }
+    if (previousAssociatedRtxStream) {
+        logger.info(primarySsrc + " was previously assocaited with rtx " +
+            previousAssociatedRtxStream + ", removing all references to it");
+        // Stream already had an rtx ssrc that is different than the one given,
+        //  remove all trace of the old one
+        videoMLine.ssrcs = videoMLine.ssrcs
+            .filter(ssrcInfo => ssrcInfo.id !== previousAssociatedRtxStream);
+        logger.info("groups before filtering for " + 
+            previousAssociatedRtxStream);
+        logger.info(JSON.stringify(videoMLine.ssrcGroups));
+        videoMLine.ssrcGroups = videoMLine.ssrcGroups
+            .filter(groupInfo => {
+                return groupInfo
+                    .ssrcs
+                    .indexOf(previousAssociatedRtxStream + "") === -1;
+            });
+    }
+    videoMLine.ssrcs.push({
+        id: rtxSsrc,
+        attribute: "cname",
+        value: primarySsrcCname
+    });
+    videoMLine.ssrcs.push({
+        id: rtxSsrc,
+        attribute: "msid",
+        value: primarySsrcMsid
+    });
+    videoMLine.ssrcGroups = videoMLine.ssrcGroups || [];
+    videoMLine.ssrcGroups.push({
+        semantics: "FID",
+        ssrcs: primarySsrc + " " + rtxSsrc
+    });
+}
+/**
+ * End helper functions
+ */
+
+/**
+ * Adds any missing RTX streams for video streams
+ *  and makes sure that they remain consistent
+ */
+export default class RtxModifier {
+    /**
+     * Constructor
+     */
+    constructor () {
+        /**
+         * Map of video ssrc to corresponding RTX
+         *  ssrc
+         */
+        this.correspondingRtxSsrcs = new Map();
+    }
+
+    /**
+     * Clear the cached map of primary video ssrcs to
+     *  their corresponding rtx ssrcs so that they will
+     *  not be used for the next call to modifyRtxSsrcs
+     */
+    clearSsrcCache () {
+        this.correspondingRtxSsrcs.clear();
+    }
+
+    /**
+     * Explicitly set the primary video ssrc -> rtx ssrc
+     *  mapping to be used in modifyRtxSsrcs
+     * @param {Map} ssrcMapping a mapping of primary video
+     *  ssrcs to their corresponding rtx ssrcs
+     */
+    setSsrcCache (ssrcMapping) {
+        logger.info("Setting ssrc cache to ", ssrcMapping);
+        this.correspondingRtxSsrcs = ssrcMapping;
+    }
+
+    /**
+     * Adds RTX ssrcs for any video ssrcs that don't
+     *  already have them.  If the video ssrc has been
+     *  seen before, and already had an RTX ssrc generated,
+     *  the same RTX ssrc will be used again.
+     * @param {string} sdpStr sdp in raw string format
+     */
+    modifyRtxSsrcs (sdpStr) {
+        let parsedSdp = transform.parse(sdpStr);
```

Comment:
hm, but this file only calls it twice?  once for each of the flows (modifyRtxSsrcs on the local side, and implodeRemoteRtxSsrcs on the remote side)